src/inventory_ui.cpp

```diff
@@ -767,10 +876,87 @@ void inventory_selector::remove_dropping_items( player &u ) const
     }
 }
 
-int inventory_selector::execute_pick( const std::string &title, const int position )
+inventory_column &inventory_selector::get_column( size_t index ) const
 {
-    prepare_paging();
-    select_item_by_position( position );
+    if( index >= columns.size() ) {
+        static inventory_column dummy;
+        return dummy;
+    }
+    return *columns[index];
+}
+
+void inventory_selector::set_active_column( size_t index )
+{
+    if( index < columns.size() && index != active_column_index && get_column( index ).activatable() ) {
+        get_active_column().on_deactivate();
+        active_column_index = index;
+        get_active_column().on_activate();
+    }
+}
+
+size_t inventory_selector::get_columns_width() const
+{
+    size_t res = 2; // padding on both sides
+
+    for( const auto &column : columns ) {
+        res += column->get_width();
+    }
+    return res;
+}
+
+void inventory_selector::toggle_active_column()
+{
+    size_t index = active_column_index;
+
+    do {
+        index = ( index + 1 < columns.size() ) ? index + 1 : 0;
+    } while( index != active_column_index && !get_column( index ).activatable() );
+
+    set_active_column( index );
+}
+
+void inventory_selector::toggle_navigation_mode()
+{
+    static const std::map<navigation_mode, navigation_mode> next_mode = {
+        { navigation_mode::ITEM, navigation_mode::CATEGORY },
+        { navigation_mode::CATEGORY, navigation_mode::ITEM }
+    };
+
+    navigation = next_mode.at( navigation );
+    for( auto &column : columns ) {
+        column->on_mode_change( navigation );
+    }
+}
+
+void inventory_selector::insert_column( decltype( columns )::iterator position, inventory_column *new_column )
+{
+    assert( new_column != nullptr );
+
+    std::unique_ptr<inventory_column> column( new_column );
```

Comment:
This looks risky. Why are you wrapping a non-unique ptr in a unique ptr? Why not pass unique one as argument instead?
Passing unique ptr would make the intention clear and prevent possible bugs resulting from someone trying to keep the column despite inserting it.

Also, all usages of the function are in the form:
```
insert_column( iter, uniqueptr.release() );
```

You could just change the `ptr.release()` into `std::move( ptr )`.